castai/resource_node_template.go

```diff
@@ -1075,9 +1123,27 @@ func toTemplateConstraints(obj map[string]any) *sdk.NodetemplatesV1TemplateConst
 			out.Spot = toPtr(!v)
 		}
 	}
-	if v, ok := obj[FieldNodeTemplateStorageOptimized].(bool); ok {
+
+	// Deprecated field, will be removed in the next major release.
+	if v, _ := obj[FieldNodeTemplateStorageOptimized].(bool); v {
 		out.StorageOptimized = toPtr(v)
 	}
+
+	// Until the next major release, we need to keep the backward compatibility with the old field set to true.
+	// otherwise, default value from the new field will be used and remove the constraint.
+	if !lo.FromPtr(out.StorageOptimized) {
```

Comment:
[note] This creates a side effect opposite of CSU.

`xx_optimized = false` is ignored and removed from the template even if it is set explicitly because we don't have a way to understand if it is set or zero value

However, this is also clearly indicated in the deprecation message.